### DISCUSSION TITLE
Update on SITL

### DISCUSSION THREAD

**Discussion by DanielAmigo** [unl-nimbus-lab/pymavswarm#130] [General]
Hi @evan-palmer !

Sorry I took so long to reply, it has been quite busy recently and couldn't make time to continue with this project.
I have progressed a bit with the SITL test. 

Right now I'm able to put drones and pymavswarm detects them, not all the times, but it is good enough. I did several tests with MavProxy, but using QGroundControl's MAVLink forwarding it is easier.
I had to specify agents to ignore (QGC I guess it is) and create a loop to wait until it detects all drones.
I only needed to overload the connect functions in mavswarm.py and connection.py to connect to mavlink with a URL, just to pass the address set on QGC: mavswarm.connect(url="udp:127.0.0.1:14551")

So this initial step is quite completed, but in the mission itself I struggle a bit. Some commands work, therefore I guess everything is working correctly about pymavlink and pymavswarm. I put mavswarm.set_mode("TAKEOFF", ...) and see in QGC how it has been changed, I put mavswarm.arm() and see the simulated drone init the propellors, but the main problem is that I ask for a simple takeoff and the UAV rejects it.
From my initial analysis it is related to pymavlink, so I'm trying to create simple missions with pymavlink directly. As I explained, we were using MavSDK-Python, which as far as I see, it is way easier than pymavlink, and maybe due to it and the differences between PX4 and Ardupilot I'm struggling to use those takeoff commands you have implemented.

So that's all, just a brief update to let you know I continue with it and hopefully soon I can figure out why PX4's reject my takeoff commands. I guess it should be easier if I had more experience with real drones, but it is what it is.

**Comment by evan-palmer** [comment]
> Sorry I took so long to reply, it has been quite busy recently and couldn't make time to continue with this project. I have progressed a bit with the SITL test.

No problem at all! The Ph.D. life is a busy one :smiley: 
 
> I had to specify agents to ignore (QGC I guess it is) and create a loop to wait until it detects all drones. I only needed to overload the connect functions in mavswarm.py and connection.py to connect to mavlink with a URL, just to pass the address set on QGC: mavswarm.connect(url="udp:127.0.0.1:14551")

That's awesome! That capability will be very useful in the future. It sounds like you are doing something similar to what I have did in my experiments as well. In my experiments, I ignore the ID of the GCS and other agents that aren't being considered. There was a period where I automatically attempted to ignore the GCS in `pymavswarm` itself, but removed that in-case there happened to be some reason to use the GCS ID pair. 

> So this initial step is quite completed, but in the mission itself I struggle a bit. Some commands work, therefore I guess everything is working correctly about pymavlink and pymavswarm. I put mavswarm.set_mode("TAKEOFF", ...) and see in QGC how it has been changed, I put mavswarm.arm() and see the simulated drone init the propellors, but the main problem is that I ask for a simple takeoff and the UAV rejects it. From my initial analysis it is related to pymavlink, so I'm trying to create simple missions with pymavlink directly. As I explained, we were using MavSDK-Python, which as far as I see, it is way easier than pymavlink, and maybe due to it and the differences between PX4 and Ardupilot I'm struggling to use those takeoff commands you have implemented.

Interesting. Would you be willing to share a code snippet that demonstrates how you are implementing the takeoff sequence? If you haven't checked it out yet, I recommend referring to the [takeoff sequence example](https://github.com/unl-nimbus-lab/pymavswarm/blob/main/examples/takeoff_sequence.py). You might be able to do something similar to the example, but replace the `GUIDED` mode with `OFFBOARD`
